\section{Introduction} \label{sec:intro}
The three-dimensional structure of DNA has recently become an active field of research,
because it influences many biological processes within living organisms.
However, the lab processes required to investigate the so-called chromosome conformation
are still comparatively expensive, which has stimulated research in predicting 
3D-structure from available data.

One such software package to predict the 3D-structure of DNA, \emph{HiC-Reg}, was conceived by Zhang et al. in 2018/2019 \cite{Zhang2018,Zhang2019}.
It is using a random forest regressor to learn DNA conformation from known DNA-DNA- and DNA-protein interactions. 
Unfortunately, HiC-Reg requires customized input data, produces custom output data
and has a high demand for computational ressources, making it inconvenient to use for scientists.

To improve on HiC-Reg's weaknesses, a previous masterproject at the University of Freiburg 
has resulted in the development of \emph{hicprediction} \cite{Bajorat2019}, 
which is basically a more efficient python implementation of HiC-Reg with standard in- and output formats.
However, its predictions are currently not on par with the ones published for HiC-Reg.
The goal of this present masterproject is thus to investigate reasons for hicprediction's underperformance
and improve the predictions.

\subsection{Motivation}
Ever since the discovery of the DNA double-helix structure \cite{Watson1953} and the central dogma of molecular biology \cite{Crick1958} in the 1950's,
the vital role of Deoxyribonucleic Acid (DNA) for the existence of all living organisms has been generally known.
Less known, however, is the fact that DNA is not only a long molecule holding genetic information, 
but also forms a spatial structure termed \emph{chromosome conformation}, 
which facilitates contacts and loops among linearly distant regions in the DNA. 
Besides the DNA-sequence, the 3D-conformation has a strong influence on important DNA-driven processes
like gene expression as well \cite{Pombo2015,Bonev2016}.

While techniques to ``read'' sequences of DNA base pairs are nowadays fast and reliable, 
methods to determine the spatial structure of DNA are less mature.
These so-called \emph{chromosome conformation capture} methods all rely on a chemical fixation 
of the DNA-structure and some post-processing to count DNA-DNA interactions for certain loci of defined size. 
The original method from 2002, 3C \cite{Dekker2002}, could only determine interactions 
between two distinct loci at a time (one vs. one), while more advanced techniques like
4C \cite{Simonis2006} (one vs. all loci) and 5C \cite{Dostie2006} (many vs. many loci) 
 have a higher performance.
The currently preferred method, Hi-C \cite{LiebermanAiden2009, Berkum2010},
is capable of determining all DNA-DNA interactions on the genome scale in a single experiment 
(all vs. all loci). 
The main output of Hi-C is the so-called contact- or interaction count matrix, 
which holds the numbers of interactions between all pairs of loci, genome-wide.
Hi-C is explained in more detail in \,\autoref{sec:intro:hic}.

Despite the progress in recent years, chromosome conformation capture methods still involve prohibitively expensive and 
tedious wet-lab processes, which limit their range of applications. 
In this regard, it must also be noted, that -- unlike the DNA sequence -- 
chromosome conformation is both dynamic and cell-specific, 
so a large number of experiments is required to get an overview over a whole organism.

Irrespective of the costs, high-resolution Hi-C datasets are already available for several 
human-, mouse- and drosophila cell lines.
Research on these datasets has revealed significant correlations between 
chromosome conformation and certain proteins bound to DNA, namely specific types of histones 
and transcription factors \cite{Bonev2016, Rao2014}.
Independent of Hi-C, the corresponding interaction sites between such proteins and DNA can be found with the well-established ChIP-Seq method,
a combination of chromatin immunoprecipitation and DNA-sequencing technologies \cite{Johnson2007, Robertson2007}, see
\autoref{sec:intro:chipseq}. 
In contrast to Hi-C, this method is fast, affordable, and a plethora of datasets for different proteins and genomes
already exists in databases.
Researchers are therefore currently investigating possibilities to predict DNA-DNA interaction sites 
from DNA-protein interaction sites. Conformation capture methods similar to Capture-C \cite{Hughes2014} could then be employed 
with a focus on the predicted interaction regions, which would be much more efficient than running Hi-C experiments on the whole genome.

Detecting hidden patterns in large sets of input data and figuring out -- potentially nonlinear -- dependencies between input and output
has recently become a domain of machine learning algorithms. 
For that reason, different machine learning techniques have been applied in the last five years -- more or less successfully -- 
to predict DNA-DNA interactions and -structure from available data \cite{Zhang2019,Pierro2017,Farre2018,Schwessinger2019,Li2019}.

During a previous masterproject by Andre Bajorat \cite{Bajorat2019}, 
a regression model based on random forests as proposed by Zhang et al. \cite{Zhang2019, Zhang2018}
has been implemented, which permits a direct estimation of contact matrices from ChIP-seq data and genomic distance.
The matrices obtained with this approach look promising, but are currently not deemed good enough for general usage.

The goal of this masterproject is therefore to investigate ideas for improving the predictions of the existing machine learning algorithm.
This includes discarding, scaling and emphasizing parts of the training sets, concatenating datasets from different cell lines 
and adding input data with higher peak density than the currently used ChIP-seq narrowPeak files.

However, before going into details, 
the underlying Hi-C- and ChIP-seq processes shall be explained in more detail,
since they are central for understanding the rest of the report.

\clearpage
\subsection{Introduction to Hi-C} \label{sec:intro:hic}
\begin{wrapfigure}[36]{r}{0.3\textwidth}
 \resizebox{0.3\textwidth}{!}{
 \small
 \import{figures/}{HiC-lab.pdf_tex}}
 \caption{Hi-C lab process}
 \label{fig:intro:HiC}
\end{wrapfigure}
The Hi-C process is targeted at investigating the three-dimensional structure of DNA
by detecting DNA-DNA interactions, 
as depicted in simplified form in \autoref{fig:intro:HiC}.

The typical input to Hi-C are about 20-25 million cells of the same type \cite{Berkum2010},
which are first crosslinked to fix DNA-DNA contacts, e.\,g. using formaldehyde, 
and then lysed to extract the DNA.
Next, the obtained DNA is cut into fragments by restriction enzymes (1),
usually HindIII or DpnII, 
and the cut ends are filled up with nucleotides partially marked by biotin (2).
Blunt ends are then joined (3) under conditions which prefer
ligations among open ends over ligations between different fragments,
usually achieved by high dilution of the fragments in solvent.
The ligated fragments are then purified and sheared into shorter pieces,
some of which contain biotinylated nucleotides and some not (4).
Those fragments which contain biotinylated nucleotides 
are then pulled down (5) and subjected to paired-end DNA-sequencing (6).
In the end, the outcome of the Hi-C lab process is a bunch of short genomic sequences, so-called reads,
which are subsequently processed in the bioinformatics part of the Hi-C protocol.

On the bioinformatics side, the reads are first mapped to the relevant reference genome,
keeping only so-called chimeric reads, where the sequence from the ``left'' end of a read
uniquely maps to one region of the reference genome and the sequence from the ``right'' end 
of the read uniquely maps to another one.
These reads are subjected to quality control, and those passing are counted as an interaction
between the two regions where both ends have been mapped.
The final outcome of a Hi-C experiment is then a (sparse) matrix with interaction counts 
between all possible pairs of regions.
The size of these regions -- or resolution of the matrix -- depends on the read coverage
of the experiment and is often set to \SI{5000}{\bp}, especially in this masterproject.

Often just a small fraction of all reads fulfill the selection criteria outlined above,
which makes Hi-C a comparatively expensive process, because several billions
of reads may be required to obtain meaningful results \cite{Rao2014}.

\clearpage
\subsection{Introduction to ChIP-seq} \label{sec:intro:chipseq}
\begin{wrapfigure}[38]{l}{0.3\textwidth}
 \resizebox{0.3\textwidth}{!}{
 \small
 \import{figures/}{chipseq_wetlab.pdf_tex}}
 \caption{ChIP-seq lab process}
 \label{fig:intro:chipseq-lab}
\end{wrapfigure}
The ChIP-seq process is a combination of Chromatin-Immunoprecipitation and DNA-sequencing, 
designed for investigating DNA-protein interactions.

The typical input to ChIP-seq is a number of cells with DNA that has proteins, histones etc. attached to
it, \autoref{fig:intro:chipseq-lab}, top. 
As a first step, all these DNA-protein contacts get fixed, 
e.\,g. by adding formaldehyde to the cells (1).
The cells are then lysed, the DNA-protein structure is extracted and cut into fragments, 
for example by sonication (2).
Next, specific antibodies are added, designed to bind only to a certain protein of interest (3).
These antibodies are additionally equipped with a tag, for example a magnetic one, so that 
the DNA-protein-antibody structures can be precipitated, while fragments without antibodies are discarded (4).
The proteins and antibodies are then removed (5), 
the DNA is purified and finally sequenced (6).
Typically, a control experiment is performed together with the ChIP-seq process, 
which comprises all steps described above, except the immunoprecipitation.

As with Hi-C and all other sequencing-based methods, the outcome of the ChIP-seq lab process is
again a bunch of short genomic sequences, which are then fed into the bioinformatics part of the pipeline.

The first data processing step for ChIP-seq, too, is quality control (QC), as with all other sequencing-based processes. 
Those reads passing QC are next mapped to an appropriate reference genome, 
to find out where the protein under investigation binds to DNA.
Therefore the number of mapped reads per genomic position is counted and compared 
to the numbers of reads stemming from the control experiment mentioned above. 
So-called ``peaks'' are then called at those positions where the number of mapped ChIP-seq reads is
statistically significant compared to the control signal. 
The resulting peaks, i.\,e. their start-, end- and peak position, 
along with peak-quality- and strand information, 
are commonly stored in so-called narrowPeak files.
These are simple text files with ten tab-separated columns \cite{UniCaliforniaSantaCruz2020}, which currently also serve as input to hicprediction.


